File: tests/typesafety/typescript/escape_wrapper.ts
// expect: passes

type HtmlText = { readonly html: string };

function escapeHtml(raw: string): HtmlText {
  return {
    html: raw.replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;"),
  };
}

function renderNote(note: HtmlText): string {
  return `<p>${note.html}</p>`;
}

export const page = renderNote(escapeHtml("<b>fragile</b>"));

File: tests/typesafety/python/escape_wrapper.py
# expect: passes
# run: yes
# title: HtmlText wraps the escaped note, and render_note accepts nothing else
# improves: escape_returns_str
from dataclasses import dataclass


@dataclass(frozen=True)
class HtmlText:
    value: str


def escape_html(raw: str) -> HtmlText:
    escaped = raw.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
    return HtmlText(escaped)


def render_note(note: HtmlText) -> str:
    return f"<p>{note.value}</p>"


page = render_note(escape_html("<b>fragile</b>"))
assert page == "<p>&lt;b&gt;fragile&lt;/b&gt;</p>"
